lib/google/gax/config.rb

```diff
@@ -0,0 +1,127 @@
+
+module Google
+  module Gax
+    ##
+    # Config is a simple DSL for creating Configuration classes.
+    #
+    # @example
+    #   require "google/gax/config"
+    #
+    #   class SampleConfig
+    #     extend Google::Gax::Config
+    #
+    #     config_attr :name,   nil,         String, nil
+    #     config_attr :active, true,        true, false
+    #     config_attr :count,  nil,         Numeric, nil
+    #     config_attr :env,    :production, String, Symbol
+    #
+    #     def initialize parent_config = nil
+    #       @parent_config = parent_config unless parent_config.nil?
+    #       yield self if block_given?
+    #     end
+    #   end
+    #
+    #   config = SampleConfig.new
+    #
+    #   config.name             #=> nil
+    #   config.name = "thor"    #=> "thor"
+    #   config.name             #=> "thor"
+    #   config.name = :thor     # ArgumentError
+    #
+    module Config
+      ##
+      # Add configuration attribute methods to the configuratin class.
+      #
+      # @param [String, Symbol] key The name of the option
+      # @param [Object, nil] initial Initial value (nil is allowed)
+      # @param [Hash] opts Validation options
+      #
+      def config_attr name, default, *valid_values, &validator
+        name = String(name).to_sym
+        raise "cannot create field named parent_config" if name == :parent_config
+        raise "method #{name} already exists" if method_defined? name
+
+        raise "validation must be provided" if validator.nil? && valid_values.empty?
+        validator ||= ->(value) { valid_values.any? { |v| v === value } }
+
+        name_ivar = "@#{name}".to_sym
+
+        if default.nil?
+          # create getter
+          define_method name do
+            value = instance_variable_get name_ivar
+            if value.nil?
+              parent = instance_variable_get :@parent_config
+              return parent.send name if parent&.respond_to? name
+            end
+            value
+          end
+          # create setter
+          define_method "#{name}=" do |new_value|
+            parent = instance_variable_get :@parent_config
+            valid_value = validator.call new_value
+            # Allow nil if parent config has the same method.
+            valid_value = true if new_value.nil? && parent&.respond_to?(name)
+            raise ArgumentError unless valid_value
+            if new_value.nil?
+              remove_instance_variable name_ivar if instance_variable_defined? name_ivar
+            else
+              instance_variable_set name_ivar, new_value
+            end
+          end
+        else
+          # create getter with default value
+          define_method name do
+            value = instance_variable_get name_ivar
+            if value.nil?
+              parent = instance_variable_get :@parent_config
+              return parent.send name if parent&.respond_to? name
+              return default
```

Comment:
This line seems to be the only difference between this case and the `default.nil?` case above. But they have the same behavior. i.e. if `default` and `value` are both nil, then nil will be returned here in this version of the method -- and in the `default.nil?` method above, nil will be returned at the end when `value` is returned. So I don't think we need two versions of the method. This bottom form will work in both cases.

Reply:
I agree this is a valid refactoring that can be made with the code as it exists now.

Reply:
This has been refactored and the two separate method definitions have been combined.